### DISCUSSION TITLE
getServerSideProps is called multiple times

### DISCUSSION THREAD

**Discussion by ceyhunkeklik** [vercel/next.js#13064] [Help]
Hi,
I've discerned this issue during development. I've been investigating for a long time but I don't find any solution yet.

As you can see in the following screenshot, I'm using getServerSideProps function to fetch my data but it is called twice. Could you help me?

![Image](https://i.hizliresim.com/JimFvs.png)

next@9.4.1
styled-components@5.1.0
redux@4.0.5

**Comment by jamesmosier** [comment]
Hi @ceyhunkeklik. Could you share the rest of your code?

**Reply by ceyhunkeklik** [reply]
Hi @jamesmosier ,
Actually it's a simple page,

blog.jsx
```
import React from "react";

import WPClient from "@api/wp";

import Layout from "@components/Layout/Layout";

import BlogView, { blogPropTypes } from "@views/Blog/BlogView";

const Blog = ({ posts }) => (
  <Layout>
    <BlogView posts={posts} />
  </Layout>
);

Blog.propTypes = blogPropTypes;

export const getServerSideProps = async () => {
  const posts = await WPClient.getPosts();

  // eslint-disable-next-line no-undef
  console.log("was here", new Date());

  return {
    props: {
      posts,
    },
  };
};

export default Blog;
```


BlogView.jsx
```
import React from "react";
import PropTypes from "prop-types";
import renderHTML from "react-html-parser";
import readingTime from "reading-time";
import { CalendarOutlined, FrownOutlined } from "@ant-design/icons";

import { useRouter } from "next/router";
import Link from "next/link";

import mediumPosts from "@api/mediumPosts.json";

import {
  BlogItemWrapper,
  BlogItemName,
  BlogItemDescription,
  BlogItemDate,
  BlogItemMetaWrapper,
  BlogWrapper,
  // BlogItemCover,
  MediumWrapper,
  NoItemsContainer,
  MediumLink,
} from "@views/Blog/BlogViewStyled";

import { toMediumDate } from "@common/util/commonUtil";

import VerticalDivider from "@components/Common/VerticalDivider";
import { PageTitle } from "@components/Layout/LayoutStyled";
import HorizontalDivider from "@components/Common/HorizontalDivider";
import generateMeta from "@common/generateMeta";
import { SITE_DESCRIPTION, SITE_IMAGE, SITE_TITLE } from "@common/appConfig";

const PostLink = ({ id, text }) => (
  <Link href="/post/[pid]" as={`/post/${id}`}>
    <a>{text}</a>
  </Link>
);

PostLink.propTypes = {
  id: PropTypes.string,
  text: PropTypes.string,
};

const Blog = ({ posts }) => {
  const router = useRouter();

  const navigate = (id) => {
    router.push(`/blog/${id}`);
  };

  const generatePost = (post, isMediumPost) => {
    const content = (
      <BlogItemWrapper
        isMediumPost={isMediumPost}
        onClick={() => {
          if (!isMediumPost) {
            navigate(post.slug);
          }
        }}
      >
        {/* <BlogItemCover
          isMediumPost={isMediumPost}
          image={post.featuredMedia.medium}
        /> */}
        <div>
          <BlogItemName>{post.title}</BlogItemName>
          {post.date && (
            <BlogItemMetaWrapper>
              <BlogItemDate>
                <CalendarOutlined />
                {toMediumDate(post.date)}
              </BlogItemDate>

              <VerticalDivider type="BULLET" />
              <span>{readingTime(post.content).text}</span>
              <VerticalDivider type="BULLET" />
              <span>
                {post.view_count &&
                  `${post.view_count} View${post.view_count > 1 ? "s" : ""}`}
              </span>
              {/* <VerticalDivider type="BULLET" />
              <span>100 views</span> */}
            </BlogItemMetaWrapper>
          )}
          <BlogItemDescription>
            {renderHTML(post.description)}
          </BlogItemDescription>
        </div>
      </BlogItemWrapper>
    );

    if (isMediumPost) {
      return (
        <MediumLink
          key={post.link}
          eventLabel={post.title}
          to={post.link}
          target="_blank"
        >
          {content}
        </MediumLink>
      );
    } else {
      return <>{content}</>;
    }
  };

  const TITLE = `Blog - ${SITE_TITLE}`;

  return (
    <>
      {generateMeta(TITLE, SITE_DESCRIPTION, SITE_IMAGE, router)}

      <BlogWrapper>
        <PageTitle>Blog</PageTitle>
        {posts?.map((post, index) => (
          <div key={post.id}>
            {generatePost(post)}
            {index + 1 < posts.length && <HorizontalDivider width="150px" />}
          </div>
        ))}
        {(!posts || !posts.length) && (
          <NoItemsContainer>
            <FrownOutlined />
            <span>There is no internal post yet.</span>
          </NoItemsContainer>
        )}
        <HorizontalDivider
          width="100%"
          title="Medium Posts"
          // color="rgba(1, 158, 116, 0.8)"
        />
        <MediumWrapper>
          {mediumPosts?.map((post) => generatePost(post, true))}
        </MediumWrapper>
      </BlogWrapper>
    </>
  );
};

export const blogPropTypes = {
  posts: PropTypes.arrayOf(PropTypes.shape()),
};

Blog.propTypes = blogPropTypes;

export default Blog;
```

**Comment by ceyhunkeklik** [comment]
Up!

**Comment by connershoop** [comment]
I've had this issue when the page makes a request for an image source, it rewrites the context.params of getServerSideProps, then reruns getServerSideProps

**Reply by jflayhart** [reply]
only happens in prod build, Chrome for me lol local dev is fine. Safari and FF are fine...

**Comment by ghost** [comment]
+1 me too

**Comment by lalosh** [comment]
+1 me too

**Comment by shabeebcoder** [comment]
+ me too

**Comment by YeeJone** [comment]
me too 
getServerSideProps

**Comment by connershoop** [comment]
I marked this as an issue in the past, also reproduced it in one of Nextjs example repos.  Apparently is wasn't important so it was moved to a discussion here: https://github.com/vercel/next.js/discussions/13566.  Also it seems like it is only happening in chrome?

**Comment by jacobhytter** [comment]
I have the same problem, (checking the output from command line) and it runs getServerSideProps two times.

(I cant post the code, it will not get formated, i dont know why)

I have tryed to remove all the "extra" code and just have the bare minimum and this still happend that getServerSideProps runs two times.

**Comment by michael-goss** [comment]
+1 same issue here
weirdly enough getServerSideProps is called correctly with the intended context, e.g. /user/1120317203 and then it's been called a few more times with some files from my /public folder as context.params: i.e. /user/favicon.ico, /user/icon16.png, /user/icon32.png. 
I could only observe this issue in Chrome. Firefox and Safari work fine

**Reply by stigkj** [reply]
Do you have dynamic routes? It seems the browser tries to load the favicon through your dynamic route. Might be loading of favicon in Chrome also looks at the for it on the current path and not just in root.

**Reply by michael-goss** [reply]
yes, that's a dynamic route: /pages/user/[id].js 
Though not all of the .png files in /public directory are loaded through this route, only some.

**Comment by ivonakov** [comment]
getServerSideProps runs twice when don't have _public/favicon.ico_
( Chrome / Safari )

**Reply by michael-goss** [reply]
I'm afraid I do have a public/favicon.ico and still see the issue

**Comment by kamami** [comment]
+1 Same issue here 
getServerSideProps is called twice in /account/orders/[orderId].js

public/favicon.ico is present

**Comment by KingMatrix1989** [comment]
Same issue here but in case that SW.js is disabled, its work fine.

**Reply by don-boldyrev** [reply]
How did you disable?

**Comment by aslam-py** [comment]
+1 Same issue here
getServerSideProps is called twice in /category/[...slug].js
Even I removed all the files from /public but still facing the same issue
@timneutkens can you help me on this?

**Comment by RutgerRauws** [comment]
Having the same issue!
Oddly enough, for a page `[id].js`, the second time `getServerSideProps` is called `context.params.id` equals `null`.

**Comment by sean-clarke** [comment]
Having the same issue: getServerSideProps is being called once with the corrent context.params.id (/pages/pagename/[id].tsx) and then a second time with context.params.id being null... Any clues?

**Comment by valerioceniccola** [comment]
Having same issue, the value is undefined the second time. 

<img width="161" alt="Schermata 2021-03-12 alle 15 23 57" src="https://user-images.githubusercontent.com/5312849/110952877-fcf81c00-8346-11eb-91f8-b39366ae4a9d.png">

**Comment by valerioceniccola** [comment]
Update Next to 10.0.9 seems to solve the issue.

**Comment by johnruane** [comment]
I get the same issue with query params being undefined on second running of getServerSideProps in my 'dynamic' route. One thing i noticed is that having an image that uses the package ReactProgressiveImage is causing the query param to be undefined.

I have a repo here if anyone wants to see the code in action: https://github.com/johnruane/dynamic-route-test

It's as if nextjs is statically optimising my dynamic page as the params are null (https://nextjs.org/docs/routing/dynamic-routes#caveats), even though it states that it doesn't if getServerSideProps is present (https://nextjs.org/docs/advanced-features/automatic-static-optimization)

Updating to 10.0.9 did not solve this issue.

**Comment by HugoRodriguesQW** [comment]
+1 me too
![1618084343192](https://user-images.githubusercontent.com/71078903/114282906-5b8fe300-9a36-11eb-8fdb-b5d39939498c.png)

- Yes, it's a dynamic route.
- NextJS version: 10.0.7
- It's running twice. One with normal parameters and the other with favicon.png

file: [[username].tsx](https://github.com/HugoRodriguesQW/next-level-week-4/blob/main/moveit-whats/src/pages/view/%5Busername%5D.tsx)

**Comment by amirgamil** [comment]
+1 same issue
I have next.js v10.1.3 and still the same issue. Weird thing is ctx changes on the second run despite it being at the same route - any updates on this issue?

**Comment by rodrigowpl** [comment]
same here, any on this?

**Comment by helloromero** [comment]
+1 same issue

**Comment by jpcmf** [comment]
+1 same issue, running "next": "10.1.3"

**Comment by clementohNZ** [comment]
+1 same issue

**Comment by rzaharenkov** [comment]
+1 same experience, `getServerSideProps` is called twice.

**Comment by paufar** [comment]
Same here, does anyone have a solution?

**Comment by timneutkens** [comment]
Try doing `curl http://localhost:3000`, you'll see that `getSeverSideProps` is not called twice (unless you have a wrongly set up custom server). It's likely that what you're seeing is happening on a catch all or dynamic route and it's matching for `/favicon.ico`, other cases could be you have a chrome extension installed that fetches a particular route in the app as well.

**Reply by johnruane** [reply]
Yes I did the curl command and confirmed getServerSideProps is only called once. I removed the favicon and disabled all react extentions but still get the issue. My route is a dynamic route. Is there anything else it could be? As mentioned if i remove react-progressive-image then I don't see this issue, but not sure what this package would be doing to cause this.

**Comment by libeyondea** [comment]
same issue

**Comment by meliborn** [comment]
In my case, it's because of the broken url to request an image. Take a look at your dev console and try to find 404 codes

**Comment by yuriy-eleks** [comment]
2021, 19, Jul issue still exists

**Comment by djleonskennedy** [comment]
Can be reproduced even in `dev` mode

https://github.com/FormidableLabs/react-progressive-image/issues/34

**Comment by Muralidevan** [comment]
Problem still exists +1

**Comment by johnruane** [comment]
@meliborn was correct in my case. I wasn't passing an image url correctly into a component i was using.

**Comment by sivadinesh1** [comment]
+1 its annoying . can the team look into this.

**Comment by bryanprimus** [comment]
+1

**Comment by timneutkens** [comment]
Please read https://github.com/vercel/next.js/discussions/13064#discussioncomment-733457
